ext/msgpack/packer_ext_registry.h

```diff
@@ -89,20 +89,20 @@ static inline VALUE msgpack_packer_ext_registry_lookup(msgpack_packer_ext_regist
      * 1. check whether singleton_class of this instance is registered (or resolved in past) or not.
      *
      * Objects of type Integer (Fixnum, Bignum), Float, Symbol and frozen
-     * String have no singleton class and raise a TypeError when trying to get
-     * it. See implementation of #singleton_class in ruby's source code:
-     * VALUE rb_singleton_class(VALUE obj);
-     *
-     * Since all but symbols are already filtered out when reaching this code
-     * only symbols are checked here.
+     * String have no singleton class.
+     * However Ruby doesn't expose any API to access the singleton_class without creating it if it
+     * doesn't exist yet. So we access the `RBasic` struct directly and check if the class is a singleton
+     * See implementation of `VALUE rb_singleton_class_get(VALUE obj)` in ruby's source code.
      */
-    if (!SYMBOL_P(instance)) {
-      lookup_class = rb_singleton_class(instance);
+    if (!SPECIAL_CONST_P(instance)) {
+      lookup_class = RBASIC(instance)->klass;
 
-      type = msgpack_packer_ext_registry_fetch(pkrg, lookup_class, ext_type_result);
+      if (FL_TEST(lookup_class, FL_SINGLETON)) {
+          type = msgpack_packer_ext_registry_fetch(pkrg, lookup_class, ext_type_result);
 
-      if(type != Qnil) {
-          return type;
+          if(type != Qnil) {
+              return type;
+          }
```

Comment:
I'll submit part of this as a separate PR, because unless I'm mistaken, this code cause all extension objects to have their `singleton_class` defined when they are serialized, which is pretty bad for performance.

IMO Ruby should expose a way to check if an object has a singleton class without creating it while doing so. I'll open a ticket upstream as well to add such feature. It already exists as `VALUE rb_singleton_class_get(VALUE obj)` but isn't exposed to extensions, and there's no Ruby side equivalent.

Reply:
Oh, you are already considering this option! I'd checked #245 at first (because it looked more specific/focused change) and noted the same idea there:
https://github.com/msgpack/msgpack-ruby/pull/245#issuecomment-1023779134